pkg/agent/secondarynetwork/ipam/ipam_delegator.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipam
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/containernetworking/cni/pkg/invoke"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"k8s.io/klog/v2"
+)
+
+const (
+	ipamGlobal     = "whereabouts"
+	defaultCNIPath = "/opt/cni/bin"
+)
+
+func GetIPAMGlobalAddress(netConfig []byte, cmdArgs *invoke.Args) (*current.Result, error) {
+	var success = false
+	defer func() {
+		if !success {
+			// Rollback to delete assigned network configuration for failed to execute Add operation
+			cmdArgs.Command = "DEL"
+			if err := delegateNoResult(ipamGlobal, netConfig, cmdArgs); err != nil {
+				klog.ErrorS(err, "Failed to roll back to delete configuration ",string(netConfig), err)
+			}
+		}
+	}()
+
+	r, err := delegateWithResult(ipamGlobal, netConfig, cmdArgs)
+	if err != nil {
+		klog.ErrorS(err, "Failed whereabouts: ", err)
+		return nil, err
+	}
+
+	ipamGlobalResult, err := current.NewResultFromResult(r)
+	if err != nil {
+		return nil, err
+	}
+	success = true
+	return ipamGlobalResult, nil
+}
+
+func DelIPAMGlobalAddress(netConfig []byte, cmdArgs *invoke.Args) error {
+	cmdArgs.Command = "DEL"
+	if err := delegateNoResult(ipamGlobal, netConfig, cmdArgs); err != nil {
+		return err
+	}
+	return nil
+}
+
+var defaultExec = &invoke.DefaultExec{
+	RawExec: &invoke.RawExec{Stderr: os.Stderr},
+}
+
+func delegateCommon(delegatePlugin string, exec invoke.Exec, cniPath string) (string, invoke.Exec, error) {
```

Comment:
Looks like there is a fair amount of code duplication with `/agent/cniserver/ipam/ipam_delegator.go`, is there an option to reuse the code?

Reply:
I agree.

Reply:
apologies for the delayed reply on this comment (though I communicate with Jianjun/Antonin over Slack on the inputs for their comments).
The reason I didn't reuse the existing implementation for ipam was due to,
1) Antonin and Jianjun suggested to keep the secondary interface changes as isolated as feasible.
2) during initial development, I thought we will have more to be done w.r.t whereabouts CNI parsing the retrieved IPAM context from networkAttachementDefinition. But, that wasn't the case.

Indeed, I see that we would be able to reuse the IPAM package with changes/modifications to the existing implementation. If this is fine, I can get it addressed asap.

Reply:
Yeah, from what I saw a few funcs here are exactly the same as those in cniserver/ipam, so probably we should reuse them.

Reply:
So, you plan to refactor this part and share code with `/agent/cniserver/ipam/ipam_delegator.go` in a later PR?

Reply:
yes @jianjuns , @annakhm I will do this as separate PR. Need to ensure that the existing IPAM code refactor will not impact primary interface IPAM use case, while it will be used for both the needs (host-local and whereabouts).